app/controllers/characters_controller.rb

```diff
@@ -83,35 +86,44 @@ def update
         @character.gallery_groups = process_tags(GalleryGroup, :character, :gallery_group_ids)
         @character.save!
       end
-      flash[:success] = "Character saved successfully."
-      redirect_to character_path(@character)
     rescue ActiveRecord::RecordInvalid
       @page_title = "Edit Character: " + @character.name
-      flash.now[:error] = {}
-      flash.now[:error][:message] = "Your character could not be saved."
-      flash.now[:error][:array] = @character.errors.full_messages
+      flash.now[:error] = {
+        message: "Your character could not be saved.",
+        array: @character.errors.full_messages
+      }
       build_editor
       render :edit
+    else
+      flash[:success] = "Character saved successfully."
+      redirect_to character_path(@character)
     end
   end
 
   def duplicate
     dupe = @character.dup
 
-    Character.transaction do
-      dupe.gallery_groups = @character.gallery_groups
-      dupe.settings = @character.settings
-      dupe.ungrouped_gallery_ids = @character.ungrouped_gallery_ids
-      @character.aliases.find_each do |calias|
-        dupalias = calias.dup
-        dupe.aliases << dupalias
-        dupalias.save!
+    begin
+      Character.transaction do
+        dupe.gallery_groups = @character.gallery_groups
+        dupe.settings = @character.settings
+        dupe.ungrouped_gallery_ids = @character.ungrouped_gallery_ids
+        @character.aliases.find_each do |calias|
+          dupalias = calias.dup
+          dupe.aliases << dupalias
+          dupalias.save!
+        end
+        dupe.save!
       end
-      dupe.save!
+    rescue ActiveRecord::RecordInvalid
+      flash[:error] = {
+        message: "Character could not be duplicated.",
+        array: dup.errors.full_messages
+      }
```

Comment:
I don't think `character#duplicate` has a view associated, and we don't redirect here, so this path should fail. I'm also not confident that `ActiveRecord::RecordInvalid` errors will get spit out of a transaction block. Could we leave this as it was – that way it'll send us an error email, if it does ever happen (and we can code something appropriately then)?

Reply:
I could do that, or I could fake up a test, just a sec

Reply:
Looks good!